text/0000-serialization.md

```diff
@@ -0,0 +1,394 @@
+- Neither FastNetworkReader nor FastNetworkWriter inherits from nor contains a `Stream`. Any code that currently operates on streams will have to be rewritten.
+- FastNetworkReader and FastNetworkWriter are heavily optimized for speed, using aggressive inlining and unsafe code to achieve the fastest possible buffer storage and retrieval.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+There are certainly some good reasons not to take this approach:
+
+- The proposed serializers use unsafe code to maximize speed (we could use NativeArray's built-in methods to gain its safety checks, but would lose considerable performance in so doing)
+- The proposed design is restrictive and necessitates coding in a style that avoids allocations
+- The proposed design places fixed limits on various values: the size of the send buffer, the number of items that may be received per frame, etc.
+- The proposed design uses fixed-size receive queues, which will use memory even when empty (though the size of the ReceiveQueueItem is relatively small, so the memory use should be reasonable)
+- The proposed design expects the user to accurately create a `GetSizeUpperBound()` method for each message type
+- The proposed design removes some control over when messages are sent, since they will be sent any time a buffer becomes full
+- The proposed design encourages (though it does not require) messages to be unmanaged types
+- The proposed design does move us a step away from jobifying message serialization (but also may in fact eliminate the need or value of jobifying message serialization)
+
+However...
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- This approach achieves extremely fast performance
+- This approach avoids memory allocations at the core layer except by way of `NativeArray`'s `Allocator.Temp` and `Allocator.TempJob`, which are orders of magnitude faster, and are only used in situations where incoming data needs to be moved from memory we don't own to memory we do, or where we need a temporary buffer to avoid reserializing.
+- This approach makes our message formats clean, easy to understand, and easy to debug.
+- This approach gives each message its own "home" in the form of the struct that contains it, making it easy to find all the code that serializes, deserializes, and processes the message.
+- This approach serializes data exactly once, removing a current step wherein data is deserialized, changed, and reserialized
+
+Ultimately, this approach not only increases the performance of MLAPI, but also reduces the barrier of entry for new developers and increases the readability and transparency of our codebase - all of which will have the effect of increasing our users' ability to feel confident in MLAPI.
+
+# Prior art
```

Comment:
There is some prior work here which could be explored (I think overall what's presented in this RFC is better than the prior work but I think it might still be interesting to take a look at it)
- [UNet NetworkMessages](https://docs.unity3d.com/Manual/UNetMessages.html)
- [Mirror NetworkMessages](https://github.com/vis2k/Mirror/blob/master/Assets/Mirror/Tests/Editor/NetworkMessageTests.cs)

Reply:
Yep, was planning on looking at those, as well as looking more at DOTS netcode, but wanted to get some eyes on the current proposal while I look at them.